### PR TITLE
feat(storage): implement new hummock uploader

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

In this PR, we supported a new hummock uploader. 
Detailed can be found at https://www.notion.so/risingwave-labs/Implementation-Details-of-new-Hummock-Uploader-4f1bac068b5f474483f68b657a3c4c8f. Reading the doc is recommended before reviewing the PR.

By the way, hummock storage v1 is totally decoupled with the event handler, and the event handler is decoupled with local version manager after this PR is merged.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

**Review by hzxa21 — COMMENTED** [review]
LGTM. The codes are very clear and well-documented. Great job!

Please add more tests before merging.

**Comment by wenym1** [comment]
> Please add more tests before merging.
@hzxa21 I have added unit tests for the uploader.

File: src/storage/src/hummock/event_handler/uploader.rs [hunk 1]
```diff
@@ -0,0 +1,1249 @@
...
+            divide_version: 0,
+        })]
+    }
+
+    fn test_uploader_context<F, Fut>(upload_fn: F) -> UploaderContext
+    where
+        Fut: UploadOutputFuture,
+        F: UploadFn<Fut>,
+    {
+        let upload_fn = Arc::new(upload_fn);
```
Comment (hzxa21) [inline src/storage/src/hummock/event_handler/uploader.rs:768]: minor: no need to wrap it in an Arc and clone in the codes below

File: src/storage/src/hummock/event_handler/uploader.rs [hunk 2]
```diff
@@ -0,0 +1,1249 @@
...
+            BufferTracker::for_test(),
+        )
+    }
+
+    fn test_uploader<F, Fut>(upload_fn: F) -> HummockUploader
+    where
+        Fut: UploadOutputFuture,
+        F: UploadFn<Fut>,
+    {
+        let upload_fn = Arc::new(upload_fn);
```
Comment (hzxa21) [inline src/storage/src/hummock/event_handler/uploader.rs:784]: minor: same here

**Review by hzxa21 — APPROVED** [review]
LGTM

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs [hunk 3]
```diff
@@ -144,99 +182,116 @@ impl HummockEventHandler {
...
-                        .inspect_err(|e| {
-                            error!("sync upload task failed: {}, err: {:?}", sync_epoch, e);
-                        });
+    fn handle_epoch_synced(
+        &mut self,
+        epoch: HummockEpoch,
+        newly_uploaded_sstables: Vec<StagingSstableInfo>,
+    ) {
+        if !newly_uploaded_sstables.is_empty() {
+            let mut read_version_guard = self.read_version.write();
```
Comment (Li0k) [inline src/storage/src/hummock/event_handler/hummock_event_handler.rs:211]: Can we prune staging_sst here ? and no need to hold the write lock of read_version and call update with not associated staging_sst  ?  Shall we add some comments and achieve in other PR

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs [hunk 4]
```diff
@@ -144,99 +182,116 @@ impl HummockEventHandler {
...
                 );
             }
         }
     }
 
-    fn handle_sync_epoch(
+    fn handle_data_spilled(&mut self, staging_sstable_info: StagingSstableInfo) {
+        self.read_version
+            .write()
+            .update(VersionUpdate::Staging(StagingData::Sst(
```
Comment (Li0k) [inline src/storage/src/hummock/event_handler/hummock_event_handler.rs:255]: minor: same here

**Review by Li0k — APPROVED** [review]
Rest LGTM, great job !!!

File: src/storage/src/hummock/event_handler/hummock_event_handler.rs (hunk 3 above)
Comment (wenym1) [reply src/storage/src/hummock/event_handler/hummock_event_handler.rs:211]: In the current code, there is only one ReadVersion, so I think there is no need to do the pruning in this PR. Maybe we can do it when we really support sharding.

**Pull request merged**